application-xpoll-api/src/main/java/com/xwiki/xpoll/internal/DefaultXPollManager.java

```diff
@@ -0,0 +1,209 @@
+
+    private static final LocalDocumentReference XPOLL_VOTES_CLASS_REFERENCE =
+        new LocalDocumentReference(XPOLL_SPACE_NAME, "XPollVoteClass");
+
+    private static final String PROPOSALS = "proposals";
+
+    private static final String VOTES = "votes";
+
+    private static final String WINNER = "winner";
+
+    private static final String USER = "user";
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    @Named("compactwiki")
+    private EntityReferenceSerializer<String> serializer;
+
+    @Override
+    public void vote(DocumentReference docReference, DocumentReference user, List<String> votedProposals)
+        throws XPollException
+    {
+        XWikiContext context = contextProvider.get();
+        XWikiDocument document = null;
+        try {
+            document = context.getWiki().getDocument(docReference, context);
+            setUserVotes(votedProposals, context, document, user);
+            updateWinner(context, document);
+        } catch (XWikiException e) {
+            throw new XPollException(String.format("Failed to vote for [%s] on behalf of [%s].", docReference, user),
+                e);
+        }
+    }
+
+    @Override
+    public String getRestURL(DocumentReference documentReference)
+    {
+
+        String contextPath = contextProvider.get().getRequest().getContextPath();
+        StringBuilder stringBuilder = new StringBuilder();
+        try {
+            stringBuilder.append(contextPath);
+            stringBuilder.append("/rest/wikis/");
+            stringBuilder
+                .append(URLEncoder.encode(documentReference.getWikiReference().getName(), XWiki.DEFAULT_ENCODING));
+            for (SpaceReference spaceReference : documentReference.getSpaceReferences()) {
+                stringBuilder.append("/spaces/");
+                stringBuilder.append(URLEncoder.encode(spaceReference.getName(), XWiki.DEFAULT_ENCODING));
+            }
+            stringBuilder.append("/pages/");
+            stringBuilder.append(URLEncoder.encode(documentReference.getName(), XWiki.DEFAULT_ENCODING));
+            stringBuilder.append("/xpoll");
+            return stringBuilder.toString();
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(String.format("Failed to retrieve the REST URL of the document: [%s]",
+                documentReference), e);
+        }
+    }
+
+    @Override
+    public Map<String, Integer> getVoteResults(DocumentReference documentReference)
+    {
+        XWikiContext context = contextProvider.get();
+        try {
+            XWikiDocument doc = context.getWiki().getDocument(documentReference, context);
+            List<BaseObject> xpollVotes = doc.getXObjects(XPOLL_VOTES_CLASS_REFERENCE);
+            BaseObject xpollObj = doc.getXObject(XPOLL_CLASS_REFERENCE);
+            List<String> proposals = xpollObj.getListValue(PROPOSALS);
```

Comment:
This can throw a ``NullPointerException`` that won't get caught by the ``catch (XWikiException e)`` below. So you have 3 options:
* ``catch (XWikiException | NullPointerException e)`` but this is a bad practice. ``NullPointerException`` is an unchecked exception which means it's not expected. In other words, you don't catch it, you prevent it.
* add an ``if`` here and make the method return the empty map at the end, either because it didn't enter the ``if`` or because there was an (expected) exception
* add an ``if`` here and throw an ``XPollException`` if the ``XPollClass`` object is missing; but if you go this path then you should throw an ``XPollException`` below also, when an ``XWikiException`` is caught, instead of returning the empty map (it's all or nothing: either you throw exceptions for all invalid use cases or you fail silently with the empty map for all invalid use cases); the script service would then have to catch the exceptions and return ``null`` or empty map.

The third option is probably the cleanest.